aws/product/awsec2product.go

```diff
@@ -0,0 +1,191 @@
+	if value, ok := value.(string); ok {
+		if value, ok := strconv.Atoi(value); ok == nil {
+			switch name {
+			case "vcpu":
+				dbAwsProduct.Vcpu = value
+			case "ecu":
+				dbAwsProduct.Ecu = value
+			}
+		}
+	}
+}
+
+// storeAttributeGenericBool stores an bool value in the struct
+// which will be imported in the database.
+func storeAttributeGenericBool(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
+	if value, ok := value.(string); ok {
+		var val int
+		if "Yes" == value {
+			val = 1
+		}
+		switch name {
+		case "currentGeneration":
+			dbAwsProduct.CurrentGeneration = val
+		}
+	}
+}
+
+// storeAttributeGenericString stores an string value in the struct
+// which will be imported in the database.
+func storeAttributeGenericString(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
+	if value, ok := value.(string); ok {
+		switch name {
+		case "instanceType":
+			dbAwsProduct.InstanceType = value
+		case "memory":
+			dbAwsProduct.Memory = value
+		case "storage":
+			dbAwsProduct.Storage = value
+		case "networkPerformance":
+			dbAwsProduct.NetworkPerformance = value
+		case "tenancy":
+			dbAwsProduct.Tenancy = value
+		case "operatingSystem":
+			dbAwsProduct.OperatingSystem = value
+		}
+	}
+}
+
+// importResult parses the body returned by downloadJSON and
+// inserts the pricing in the database.
+func importResult(reader io.ReadCloser, logger jsonlog.Logger) {
+	defer reader.Close()
+	body, err := ioutil.ReadAll(reader)
+	if err != nil {
+		logger.Error("Body reading failed", err)
+	}
+	var bodyMap map[string]interface{}
+	if err = json.Unmarshal(body, &bodyMap); err != nil {
+		logger.Error("JSON Unmarshal failed", err)
+		return
+	}
+	for _, product := range bodyMap["products"].(map[string]interface{}) {
+		var dbAwsProduct models.AwsProductEc2
+		product := product.(map[string]interface{})
+		dbAwsProduct.Sku = product["sku"].(string)
+		attributes := product["attributes"].(map[string]interface{})
+		for name, attribute := range attributes {
+			if fct, ok := storeAttribute[name]; ok {
+				fct(name, attribute, &dbAwsProduct, logger)
+			}
+		}
+		dbAwsProduct.Insert(db.Db)
+	}
+}
+
+// purgeCurrent purges the current pricing from the database.
+func purgeCurrent() {
+	// sql query
+	const sqlstr = `DELETE FROM trackit.aws_product_ec2`
+
+	// run query
+	models.XOLog(sqlstr)
+	db.Db.Exec(sqlstr)
+}
+
+// downloadJSON requests AWS' API and returns the body after
+// checking if there's already the same version in the database.
+func downloadJSON(logger jsonlog.Logger) (io.ReadCloser, error) {
+	hc := http.Client{}
+	req, err := http.NewRequest("GET", urlEC2Pricing, nil)
+	if err != nil {
+		return nil, err
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
+	if err != nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
```

Comment:
It’s good that you make sure we don’t process something we already have processed, though perhaps you should extract this in its own function. This one is very long.